Fix broadcast issue

bshall opened the pull request:
Fix for #191

c42f commented:
I think we can get away with something a bit simpler than this.  Here's what I had:

```julia
@inline broadcast_sizes(a::StaticArray, as...) = (Size(a), broadcast_sizes(as...)...)
@inline broadcast_sizes(a::Number, as...) = (Size(), broadcast_sizes(as...)...)
@inline broadcast_sizes() = ()
```

bshall commented:
Yeah, that is better.

c42f commented:
The failed test here isn't your fault, I restarted it.

Do you want to finish this one off by cleaning up a little and adding tests?

bshall commented:
Yeah, I'm  happy to do that. I'll make your change and add some tests.

c42f commented:
Awesome, thanks a lot! I had started messing with tests, but hadn't done much.

bshall commented:
Okay, so I've added a bunch of tests and found some issues/inconsistencies with broadcast from Base. I've listed those here #197, #198, #199 and #200 and I've added (commented out) test cases for them. Let me know if you'd like me to add anything else.

Oh, I also made a small change to the `broadcast!` function so it returns `dest` as well (just to be consistent).

c42f commented on test/broadcast.jl:
```diff
@@ -0,0 +1,157 @@
+        @test broadcast_sizes(ones(t), 1, 1) === (Size(t), Size(), Size())
+        @test broadcast_sizes(1, 1, ones(t)) === (Size(), Size(), Size(t))
+    end
+    @test broadcast((a,b,c)->0, SVector(1,1), 0, 0) == SVector(0, 0)
```
Could do with a comment `See #191` perhaps?

c42f commented on test/broadcast.jl:
```diff
@@ -0,0 +1,157 @@
+    @testset "2x2 StaticMatrix with 1x2 StaticMatrix" begin
+        m1 = @SMatrix [1 2; 3 4]
+        m2 = @SMatrix [1 4]
+        # @test @inferred(broadcast(+, m1, m2)) === @SMatrix [2 6; 4 8] #197
```
All the places where tests are currently commented out (because they're broken) you can use `@test_broken` instead.  This runs the test and verifies that it is, indeed, broken - so we won't forget to add them back in later when these issues are fixed.

bshall commented:
Great. Thanks for the feedback. I think I've made all the changes you suggested.
